sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -107,6 +108,17 @@ object UnionPushdown extends Rule[LogicalPlan] {
   }
 }
 
+object LimitPushDown extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // Push down limit when the child is project on limit
+    case Limit(expr, Project(projectList, l: Limit)) =>
+      Project(projectList, Limit(expr, l))
+
+    // Push down limit when the child is project on sort
+    case Limit(expr, Project(projectList, s: Sort)) =>
+      Project(projectList, Limit(expr, s))
+  }
```

Comment:
This is for SPARK-7289. As we discussed in https://github.com/apache/spark/pull/5821, we need an optimize rule like this.
cc @adrian-wang